File: spec/cpp_stl/test_if_struct.cpp
#include <boost/test/unit_test.hpp>

#include <if_struct.h>

#include <iostream>
#include <fstream>

BOOST_AUTO_TEST_CASE(test_if_struct) {
    std::ifstream ifs("src/if_struct.bin", std::ifstream::binary);
    kaitai::kstream ks(&ifs);
    if_struct_t* r = new if_struct_t(&ks);

    BOOST_CHECK_EQUAL(r->op1()->opcode(), 0x53);
    BOOST_CHECK_EQUAL(r->op1()->arg_str()->str(), "foo");

    BOOST_CHECK_EQUAL(r->op2()->opcode(), 0x54);
    BOOST_CHECK_EQUAL(r->op2()->arg_tuple()->num1(), 0x42);
    BOOST_CHECK_EQUAL(r->op2()->arg_tuple()->num2(), 0x43);

    BOOST_CHECK_EQUAL(r->op3()->opcode(), 0x53);
    BOOST_CHECK_EQUAL(r->op3()->arg_str()->str(), "bar");

    delete r;
}

File: spec/python/spec/test_if_struct.py
import unittest
from testformats.if_struct import IfStruct

class TestIfStruct(unittest.TestCase):
    def test_if_struct(self):
        with IfStruct.from_file('src/if_struct.bin') as r:
            self.assertEqual(r.op1.opcode, 83)
            self.assertFalse(hasattr(r.op1, 'arg_tuple'))
            self.assertEqual(r.op1.arg_str.str, u"foo")
            self.assertEqual(r.op2.opcode, 84)
            self.assertEqual(r.op2.arg_tuple.num1, 66)
            self.assertEqual(r.op2.arg_tuple.num2, 67)
            self.assertFalse(hasattr(r.op2, 'arg_str'))
            self.assertEqual(r.op3.opcode, 83)
            self.assertFalse(hasattr(r.op3, 'arg_tuple'))
            self.assertEqual(r.op3.arg_str.str, u"bar")
